### PR TITLE
Added ability for long Folder names in batocera/retrobat

### PR DESCRIPTION
**Pull request by Dreded**
#99 
not sure what else this would work in, made sure it works as well as previous code base for existing things but for batocera it will allow long Folder names so they can be scraped easily

Tested extensively with Mario Teaches Typing and C&C Red Alert

Also made it so that it will rename an already existing folder if it exists so Genre/game.pc would be renamed to game.pc if you removed the genre checkbox

and finally it will update the path in Gamelist.xml rather than creating duplicates.

### CODE REVIEW

**Comment by Dreded** [comment]
given that the request was specifically for RetroBat I will quickly modify the code as it obviously uses .pc folders too

**Comment by Dreded** [comment]
and im done now.. feel free to merge

**Comment by Dreded** [comment]
to be clear the warning is for multiple game folders of the same game.

**Comment by Voljega** [comment]
Thank you for the MR, I'll have a look and test it as soon as I can (pretty exhausted)

One thing I didn't understand (had a very very quick look) is does it work post-conversion or during the conversion ?

**Comment by Dreded** [comment]
> Thank you for the MR, I'll have a look and test it as soon as I can (pretty exhausted)
> 
> One thing I didn't understand (had a very very quick look) is does it work post-conversion or during the conversion ?

during the conversion, 
if long names checkbox is checked then it uses those instead of shortnames for the gameDir

File: util.py [hunk 1]
```diff
@@ -298,3 +299,58 @@ def buildCache(scriptDir, collectionDir, collection, logger):
     logger.log("gameplayPicCache: %i entities" % len(gameplayPicCache.keys()))
 
     return frontPicCache, titlePicCache, gameplayPicCache
+
```
Comment (Voljega) [inline util.py:302]: With the converter default beeing of not overwriting a game if it's found, I think it would be better to make this detection and handling a full separate functionnality in another PR, associated to a parameter.

Moving the game would likely break some of them too

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+certifi==2021.10.8
+chardet==4.0.0
+charset-normalizer==2.0.12
+idna==3.3
+Pillow==9.0.1
+requests==2.27.1
+tk==0.1.0
+urllib3==1.26.8
```
Comment (Voljega) [inline requirements.txt:8]: is this file needed and by what kind of dependencies manager ?

File: metadatahandler.py [hunk 3]
```diff
@@ -122,6 +122,9 @@ def __writeGamelistEntry__(self, gamelist, dosGame, game, genre, useGenreSubFold
         if conversionType == util.retropie:
             path = "./" + genre + "/" + util.getCleanGameID(dosGame,'.conf') if useGenreSubFolders \
                 else "./" + util.getCleanGameID(dosGame, '.conf')
+        elif (conversionType == util.batocera or conversionType == util.retrobat) and useLongFolderNames:
```
Comment (Voljega) [inline metadatahandler.py:125]: here we can test for useLongFolderNames only as we will enforce that it is always false in the GUI before starting generation when we are not using batocera or retrobat conversion

File: gamegenerator.py [hunk 4]
```diff
@@ -135,8 +137,8 @@ def __postConversionForEmuelec__(self):
         shutil.move(os.path.join(self.getLocalGameOutputDir()), emuElecDataDir)
         os.rename(os.path.join(emuElecDataDir, self.game + '.pc'), os.path.join(emuElecDataDir, self.game))
         # move *.bat *.map and *.cfg to pc/*.pc folder and rename *.cfg to dosbox-SDL2.conf
-        emuelecConfOutputDir = os.path.join(self.outputDir, 'pc', self.genre, self.game + ".pc") \
-            if self.useGenreSubFolders else os.path.join(self.outputDir, 'pc', self.game + ".pc")
+        emuelecConfOutputDir = os.path.join(self.outputDir, 'pc', self.genre, self.gameDir) \
```
Comment (Voljega) [inline gamegenerator.py:140]: this should never be used for emuelec so this is not needed

File: gamegenerator.py [hunk 5]
```diff
@@ -300,7 +302,7 @@ def __postConversionForRetropie__(self):
         dosboxCfg = open(os.path.join(self.getLocalGameOutputDir(), "dosbox.cfg"), 'a')
         # add mount c at end of dosbox.cfg
         romsFolder = util.getRomsFolderPrefix(self.conversionType, self.conversionConf)
-        retropieGameDir = romsFolder + "/" + self.genre + "/" + self.game + ".pc" if self.useGenreSubFolders else romsFolder + "/" + self.game + ".pc"
+        retropieGameDir = romsFolder + "/" + self.genre + "/" + self.gameDir if self.useGenreSubFolders else romsFolder + "/" + self.gameDir
```
Comment (Voljega) [inline gamegenerator.py:305]: this should never be used for retropie so this is not needed

File: exogui.py [hunk 6]
```diff
@@ -719,6 +729,7 @@ def __clickProceed__(self):
         self.logger.log('\n<--------- Starting ' + self.setKey + ' Process --------->')
         collectionDir = self.guiVars['collectionDir'].get()
         conversionType = self.guiVars['conversionType'].get()
+        useLongFolderNames = True if self.guiVars['longGameFolder'].get() == 1 else False
```
Comment (Voljega) [inline exogui.py:732]: here we should force it to False if we are not in the right type of conversion.
something like
`useLongFolderNames = True if self.guiVars['longGameFolder'].get() == 1 and (conversionType == 'batocera' || conversionType == 'retrobat' # else False`

File: exoconverter.py [hunk 7]
```diff
@@ -112,6 +117,7 @@ def __convertGame__(self, game, gamelist, totalSize, count):
         else:
             self.logger.log("  already converted in output folder")
 
+        util.checkMultipleofSameGame(self.useGenreSubFolders, metadata, genre, game, gameDir, self.outputDir, self.logger)
```
Comment (Voljega) [inline exoconverter.py:120]: better in separate PR as a full fledged functionnality

File: requirements.txt (hunk 2 above)
Comment (Dreded) [reply requirements.txt:8]: > is this file needed and by what kind of dependencies manager ?

That is the standard way to store python dependencies, its used by pip, those are all the packages your project needs to be installed in order to run on any system, the versions are the current versions as of when I wrote that file.

https://www.idkrtm.com/what-is-the-python-requirements-txt/

File: gamegenerator.py (hunk 4 above)
Comment (Dreded) [reply gamegenerator.py:140]: > this should never be used for emuelec so this is not needed

only for consistency

File: gamegenerator.py (hunk 5 above)
Comment (Dreded) [reply gamegenerator.py:305]: > this should never be used for retropie so this is not needed

only for consistency
